### PR TITLE
mxnet: add fp16 compression

### PR DESCRIPTION
**Pull request by vycezhong**
since horovod for mxnet supports allreduce for temporary tensors https://github.com/horovod/horovod/pull/1639 , a simple wrapper like this is sufficient.  

```python
    for i, param in enumerate(self._params):
            if param.grad_req != 'null':
                compressed, ctx = self._compression.compress(param.list_grad()[0])
                allreduce_(compressed, average=False,
                           name=param.name, priority=-i)
                param._grad[0] = self._compression.decompress(compressed, ctx)
```

### CODE REVIEW

**Comment by tgaddair** [comment]
This is awesome!  Thanks for the PR @vycezhong!  @apeforest @romerojosh @yuxihu can one of you review?

File: horovod/mxnet/compression.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        """Upcasts the tensor to the initialization dtype."""
+        tensor_decompressed = tensor
+        dtype = ctx
+        if 'float' in str(dtype):
+            tensor_decompressed = tensor.astype(dtype, copy=False)
+        return tensor_decompressed
+
+
+class Compression(object):
+    """Optional gradient compression algorithm used during push_pull."""
```
Comment (romerojosh) [inline horovod/mxnet/compression.py:69]: Change `push_pull` to `allreduce` here to be consistent with Horovod terminology.

**Comment by apeforest** [comment]
> Implementation looks good to me, just left a minor comment. Should this feature also be implemented for `DistributedOptimizer`? @apeforest, any thoughts on this?

I guess for now DistributedTraining is fine. Gluon API is going to be the default training API in MXNet going forward.

**Comment by Adnios** [comment]
How about this pr? It seems that Horovod has not yet added compression for mxnet

**Pull request closed without merging**